Graphics and Graphic Displays Articles (#178)
xodio/xod-docs: docs/guide/graphics-images/README.md

## Image node

To render a bitmap node, use the `image` node.

![Image node](./image-node.png)

The `image` node defines a rectangular shape area on a [canvas](/docs/guide/graphics-library#canvas). During the rendering process, this area fills up with the specified bitmap image. The area is set on canvas by a pivot point (`X`, `Y`), which is its upper-left vertex, width `W`, and height `H`. If the actual geometric dimensions of the image stored in the bitmap array are larger than the specified area, then only a part of the image is displayed on the screen. This area is tiled with the original image if the specified area is larger than the image size.

By `GFX` pins, the `image` node takes a part of the tree of graphic elements of the scene as well as other graphic nodes in the [graphics library](/docs/guide/graphics-library). A `BMP` pin of the `Bitmap` type. This pin links the `image` node to a specific bitmap that should be rendered.

## Embedded icons set

> The [graphics library](/docs/guide/graphics-library) has a small set of 15 example icons to get acquainted with the bitmaps.

Reviewer: ```suggestion
The [graphics library](/docs/guide/graphics-library) has a small set of example icons to get acquainted with the bitmaps.
```
This might change.

Ten 16x16 color icons, five of them with the `colorDepth` equal `2` and `0x0000` color mask:

![Color icons](./rgb-icons.jpg)
Снять на норм фотик!

- The "Heart" icon. `heart-16x16-rgb` and `heart-16x16-rgba` with mask color;
- The "Temperature" icon. `temperature-16x16-rgb` and `temperature-16x16-rgba` with mask color;
- The "Humidity" icon. `humidity-16x16-rgb` and `humidity-16x16-rgba` with mask color;
- The "Warning" icon. `warning-16x16-rgb` and `warning-16x16-rgba` with mask color;
- The "Wi-Fi" icon. `wifi-16x16-rgb` and `wifi-16x16-rgba` with mask color.

Five 16x16 black and white icons: 

![BW icons](./bw-icons.jpg)
Снять на норм фотик!

- The "Heart" icon. `heart-16x16-bw`;
- The "Temperature" icon. `temperature-16x16-bw`;
- The "Humidity" icon. `humidity-16x16-bw`;
- The "Warning" icon. `warning-16x16-bw`;
- The "Wi-Fi" icon. `wifi-16x16-bw`.

You can study the structure of these icons to create your images.

## Making own bitmap node

How to create your own `Bitmap` node for an icon or image? Let's assume that you want to make a node containing the 100x100 pixels image of a cat. 

> First of all, make a new patch for the image and name it, for example `cat`. Put the `not-implemented-in-xod` node and `output-bitmap` node from the `xod/graphics` library onto the patch. Edit the `not-implemented-in-xod` node. The C++ code inside should look like the one listed [here](/docs/guide/graphics-images/#bitmap-node) above.

Reviewer: No links with “here”/“there” anchors, please.

```suggestion
First of all, make a new patch for the image and name it, for example `cat`. Put the `not-implemented-in-xod` node and `output-bitmap` node from the `xod/graphics` library onto the patch. Edit the `not-implemented-in-xod` node. Use the C++ [template code](#bitmap-node) above as the starting point.
```

};
```

Then set the necessary parameters in the bitmap constructor. The first parameter, a pointer to the array, can be left untouched. The second parameter is `colorDepth`. We have a color image without a color mask, so choose `1`. The third and fourth parameter defines bitmap width - `100` pixels and height - `100` pixels. We can skip the fifth parameter because we don't use the color mask.

```cpp
void evaluate(Context ctx) {
    auto state = getState(ctx);
    state->myBitmap = new Bitmap(bitmap, 1, 100, 100);
    emitValue<output_OUT>(ctx, state->myBitmap);
}
```

> That it! The node with the image is ready. However, please do not rush to display it. The image has a too high resolution for microcontrollers, and it is worth calculating how much space it takes up in the controller's flash memory. Find out the total number of pixels and multiply it by the number of bytes of color per pixel 100 * 100 * 2, and you get 20 Kb of flash memory. It is a huge chunk, and you can't display such an image on the Arduino UNO. But for high capacity controllers as Arduino MEGA or ESP8266 controllers, it is suitable.

Reviewer: Typography

```suggestion
That it! The node with the image is ready. However, please do not rush to display it. The image has a too high resolution for microcontrollers, and it is worth calculating how much space it takes up in the controller's flash memory. Find out the total number of pixels and multiply it by the number of bytes of color per pixel 100×100×2, and you get 20 KB of flash memory. It is a huge chunk, and you can't display such an image on the Arduino Uno. But for high capacity controllers as Arduino Mega 2560 or ESP8266 controllers, it is suitable.
```

- The "Warning" icon. `warning-16x16-rgb` and `warning-16x16-rgba` with mask color;
- The "Wi-Fi" icon. `wifi-16x16-rgb` and `wifi-16x16-rgba` with mask color.

Five 16x16 black and white icons: 

![BW icons](./bw-icons.jpg)

- The "Heart" icon. `heart-16x16-bw`;
- The "Temperature" icon. `temperature-16x16-bw`;
- The "Humidity" icon. `humidity-16x16-bw`;
- The "Warning" icon. `warning-16x16-bw`;
- The "Wi-Fi" icon. `wifi-16x16-bw`.

You can study the structure of these icons to create your images.

## Making own bitmap node

How to create your own `Bitmap` node for an icon or image? Let's assume that you want to make a node containing the 100x100 pixels image of a cat. 

First of all, make a new patch for the image and name it, for example `cat`. Put the `not-implemented-in-xod` node and `output-bitmap` node from the `xod/graphics` library onto the patch. Edit the `not-implemented-in-xod` node. Use the C++ [template code](#bitmap-node) above as the starting point.

Find the source image and resize it to dimensions you need using a computer. In this example, the source image is a `.JPG` file.

![Source image Cat](./cat.jpg)

> Now you need to convert the source image file to a byte array of pixel data. It is tough to do this manually, so many ready-made online web services are offering such features. For example [littlevgl.com](https://littlevgl.com/image-to-c-array) or [digole.com](https://www.digole.com/tools/PicturetoC_Hex_converter.php). Graphics library operates only color images with a resolution of 2 bytes per pixel (565 color), so select the correct conversion format.

Reviewer 2: "Graphics library operates only color images with a resolution of 2 bytes per pixel (565 color), so select the correct conversion format."
It sounds like the library works only with color images, but it also works with B&W images, which has a byte per 8 pixels...

};
```

Then set the necessary parameters in the bitmap constructor. The first parameter, a pointer to the array, can be left untouched. The second parameter is `colorDepth`. We have a color image without a color mask, so choose `1`. The third and fourth parameter defines bitmap width - `100` pixels and height - `100` pixels. We can skip the fifth parameter because we don't use the color mask.

```cpp
void evaluate(Context ctx) {
    auto state = getState(ctx);

>     state->myBitmap = new Bitmap(bitmap, 1, 100, 100);

Reviewer 2: The same thing with the `...new (state->mem) Bitmap...` here

    emitValue<output_OUT>(ctx, state->myBitmap);
}
```

> That it! The node with the image is ready. However, please do not rush to display it. The image has a too high resolution for microcontrollers, and it is worth calculating how much space it takes up in the controller's flash memory. Find out the total number of pixels and multiply it by the number of bytes of color per pixel 100×100×2, and you get 20 KB of flash memory. It is a huge chunk, and you can't display such an image on the Arduino Uno. But for high capacity controllers as Arduino Mega 2560 or ESP8266 controllers, it is suitable.

Reviewer 2: I think that the information about memory better to wrap into a "Note" block.

> Let's display the created image to the [ST7735](/docs/guide/st7735-display) screen. For this, create a new patch and put the `st7735-128x160-g-device` and `render` nodes from the `xod-dev/st7735-display`library. Then make a tree of graphic elements. Put the `canvas` and `image` nodes from the `xod/graphics` library. Set up canvas width to `128` pixels and height to `160` pixels to fit the device screen. Link the created bitmap `cat` node with the `BMP` pin of the `image` node. Set the width and height of the `image` field to `100` pixels. Put the `image` in the middle of the `canvas`. For this, set the coordinates of the upper-left corner of the image to (`14`, `30`).

Reviewer 2: It sounds like it should be a part of the first article (it has an introduction in primitives, render node, but have no information about devices). However, this article may have no information about the device node, because it's an extension for the first article.

![Cat patch](./cat-patch.png)

Upload the patch and look at screen.

![Cat screen](./cat-screen.jpg)

> Now let's try to make and display the image with a color mask. As a source file, we use the color Arduino logo 100x100 pixels.

Reviewer 2: I think this section is too long. What do you think about splitting it into two?
- Making own bitmap node
- Color masking

    auto state = getState(ctx);
    state->myBitmap = new Bitmap(bitmap, 2, 100, 100, 0xF800);
    emitValue<output_OUT>(ctx, state->myBitmap);
}
```

If we want to make an image with a mask, we set the `colorDepth` value to `2`. At the last constructor parameter, we specify the two-byte `keyColor` of the mask, which is `0xF800`. This color corresponds to the pure red RGB color `#FF0000`.

Let's expand our previous patch to display images on the [ST7735](/docs/guide/st7735-display) screen.

Change the `BG` background color of the scene on the `canvas` node. For example, to the `#FF00FF` pink. It gives us a better view of the outline of the masked image on the screen. Add the `logo` image node onto the patch. Also, add the `button` and [generic](/docs/guide/generics) `if-else` node. Link the nodes, as shown below. 

![Cat and logo patch](./cat-and-logo-patch.png)

By clicking on the `button`, we are going to change the image in the center of the screen. From the `cat` image to the `logo` with a color mask. Upload the patch to the microcontroller and take a look at the result.

<video controls autoplay muted loop>
    <source src="./manage-images-with-button.mp4" type="video/mp4">

> </video>

Reviewer 2: I feel like there should be a conclusion section 🤔 

Something like:
- You can create and render images, but ensure that your MCU has enough memory for that
- Use B&W images to optimize memory usage on the low memory microcontrollers
- Images can have a color mask
